pkg/internal/cluster/create/actions/kubeadmjoin/join.go

```diff
@@ -55,58 +55,45 @@ func (a *Action) Execute(ctx *actions.ActionContext) error {
 	if err != nil {
 		return err
 	}
-	if len(secondaryControlPlanes) > 0 {
-		if err := joinSecondaryControlPlanes(
-			ctx, allNodes, secondaryControlPlanes,
-		); err != nil {
-			return err
-		}
-	}
 
 	// then join worker nodes if any
 	workers, err := nodes.SelectNodesByRole(allNodes, constants.WorkerNodeRoleValue)
 	if err != nil {
 		return err
 	}
-	if len(workers) > 0 {
-		if err := joinWorkers(ctx, workers); err != nil {
+	if len(secondaryControlPlanes) > 0 || len(workers) > 0 {
+		if err := joinNodes(ctx, allNodes, secondaryControlPlanes, workers); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
-func joinSecondaryControlPlanes(
-	ctx *actions.ActionContext,
+func joinNodes(ctx *actions.ActionContext,
 	allNodes []nodes.Node,
 	secondaryControlPlanes []nodes.Node,
+	workers []nodes.Node,
 ) error {
-	ctx.Status.Start("Joining more control-plane nodes 🎮")
+	startMsg := "Joining more "
+	if len(secondaryControlPlanes) > 0 && len(workers) > 0 {
+		startMsg += "control-plane and worker nodes 🎮🚜"
+	} else if len(secondaryControlPlanes) > 0 && len(workers) == 0 {
+		startMsg += "control-plane nodes 🎮"
+	} else {
+		startMsg += "worker nodes 🚜"
+	}
+	ctx.Status.Start(startMsg)
 	defer ctx.Status.End(false)
 
-	// TODO(bentheelder): it's too bad we can't do this concurrently
-	// (this is not safe currently)
```

Comment:
this is only safe on recent versions, right?

Reply:
at least this change is needed, which was added in 1.15.1
https://github.com/kubernetes/kubernetes/commit/bc74ac3eaad642bc748c459b5776701f50c18094#diff-5e4c5bba67c635568de70e0c424882c6

Reply:
we should probably ship this as being concurrent only for >= 1.15.1 then ?

Reply:
yes, i guess there has to be a version check and the old way of serial join has to be supported too.